admin/app/controllers/authenticated/certifications/certification/details.js

```diff
@@ -0,0 +1,91 @@
+import { action } from '@ember/object';
+import { inject as service } from '@ember/service';
+import { alias } from '@ember/object/computed';
+import Controller from '@ember/controller';
+import { schedule } from '@ember/runloop';
+
+export default class CertificationDetailsController extends Controller {  // Properties
+  juryRate = false;
+
+  juryScore = false;
+  requestedId = '';
+
+  // Private properties
+  @service('mark-store')
+  _markStore;
+
+  // Aliases
+  @alias('details.percentageCorrectAnswers')
+  rate;
+
+  @alias('details.totalScore')
+  score;
+
+  @alias('model')
+  details;
+
+  @action
+  onUpdateRate() {
+    const competences = this.get('details.competences');
+    const { good, count, jury } = _getCertificationResultsAfterJuryUpdate(competences);
+
+    if (jury) {
+      this.set('juryRate', Math.round(good * 10000 / count) / 100);
+    } else {
+      this.set('juryRate', false);
+    }
+    // TODO: find a better way
+    schedule('afterRender', this, () => {
+      const score = this.score;
+      const competences = this.get('details.competences');
+      const newScore = competences.reduce((value, competence) => {
+        value += (typeof competence.juryScore !== 'undefined' && competence.juryScore !== false) ? competence.juryScore : competence.obtainedScore;
+        return value;
+      }, 0);
+      if (newScore !== score) {
+        this.set('juryScore', newScore);
+      } else {
+        this.set('juryScore', false);
+      }
+    });
+  }
+
+  @action
+  onStoreMarks() {
+    this._markStore.storeState({
+      score: (this.juryScore === false) ? this.score : this.juryScore,
+      marks: this.get('details.competences').reduce((marks, competence) => {
+        marks[competence.index] = {
+          level: (competence.juryLevel === false) ? competence.obtainedLevel : competence.juryLevel,
+          score: (competence.juryScore === false) ? competence.obtainedScore : competence.juryScore
+        };
+        return marks;
+      }, {})
+    });
+    this.transitionToRoute('authenticated.certifications.certification.informations', this.get('details.id'));
+  }
+}
+
+function _getCertificationResultsAfterJuryUpdate(competences) {
+  return competences.reduce((data, competence) => {
+    if (competence.answers) {
+      competence.answers.forEach((answer) => {
+        if (answer.jury) {
+          if (answer.jury === 'ok') {
+            data.good++;
+          }
+          if (answer.jury !== 'skip') {
+            data.count++;
+          }
+          data.jury = true;
+        } else {
+          data.count++;
+          if (answer.result === 'ok') {
+            data.good++;
+          }
+        }
+      });
+    }
+    return data;
+  }, { count: 0, good: 0, jury: false });
+}
```

Comment:
J'ai un peu de mal à comprendre ce que fait ce code alors j'ai essayé de le refacto. J'ai : 

- enlever un niveau d'imbrication
- supprimer les effets de bords sur le paramètre de la callback du premier reduce 

C'est normal que data.jury passe à `true` dés qu'on a une `answer.jury` ?

```suggestion
function _getCertificationResultsAfterJuryUpdate(competences) {
  return competences.reduce((data, competence) => {
    if (competence.answers) {
      return competence.answers.reduce((acc, answer) => {
        if (!answer.jury) {
          return {
            ...acc,
            count: acc.count + 1,
            good: (answer.result === 'ok') ? acc.good + 1 : acc.good,
          };
        }
        return {
          ...acc,
          jury: true,
          good: (answer.jury === 'ok') ? acc.good + 1 : acc.good,
          count: (answer.jury !== 'skip') ? acc.count + 1 :  acc.count,
        }
      }, data);
    }
    return data;
  }, { count: 0, good: 0, jury: false });
}

```

Reply:
Je garde aussi ce commentaire pour la PR de refacto :) 
Je comprends pas le return à la ligne 94 par contre ?

Reply:
Je vois pas de ligne 94 dans le snippet juste au dessus. Tu peux préciser ?

Reply:
98 pardon `return competence.answers.reduce((acc, answer) => {`

Reply:
C'est pour éviter de faire des effets de bords sur `data`. On retourne `data` à jour vu que c'est l'accumulateur de notre premier reduce qu'on a donné en entrer du second reduce.

Reply:
Ces `reduce` déguisent du code impératif en fonctionnel. Je préfère le style fonctionnel par défaut quand c'est pertinent mais si la logique est impérative, mieux vaut utiliser la syntaxe impérative du langage (`let`)  pour plus de clarté :

```js
function _getCertificationResultsAfterJuryUpdate(competences) {
  let good = 0, count = 0, jury = false;
  _.flatMap(competences, 'answers').forEach((answer) => {
    if (answer.jury) {
      if (answer.jury === 'ok') {
        good++;
      }
      if (answer.jury !== 'skip') {
        count++;
      }
      jury = true;
    } else {
      count++;
      if (answer.result === 'ok') {
        good++;
      }
    }
  });
  return { good, count, jury };
}
```

Reply:
Merci pour toutes les explications @frinyvonnick  ! 
Jeremy conseille de ne pas trop toucher ces parties de code (pas trop testées) pour le moment du coup je ne pense pas implémenter dans cette PR certaines suggestions. Mais je garde en tête tes suggestions et à l'occasion je reprendrais tes commentaires :))